src/operator/tensor/elemwise_unary_op.h

```diff
@@ -9,124 +9,335 @@
+    CHECK_EQ(src_blob.type_flag_, dest_blob.type_flag_);
+    CHECK_EQ(src_blob.shape_, dest_blob.shape_);
+    MSHADOW_TYPE_SWITCH(src_blob.type_flag_, DType, {
+      mshadow::Copy(dest_blob.FlatTo1D<xpu, DType>(s), src_blob.FlatTo1D<xpu, DType>(s));
+    });
+  }
 
-template<typename GRAD_OP>
-struct unary_bwd {
-  template<typename DType>
-  MSHADOW_XINLINE static DType Map(DType a, DType b) {
-    return DType(a*GRAD_OP::Map(b));
+  /*! \brief Allocate geometry-related blob data for sparse tensors */
+  static void AllocateGeometry(const NDArray *dest, const NDArray* clone_from = nullptr) {
+    if (clone_from) {
+      const TShape ishape = clone_from->storage_shape();
+      TShape sshape = dest->storage_shape();
+      CHECK(shape_assign(&sshape, ishape));
+      dest->CheckAndAllocData(sshape);
+      CHECK_EQ(dest->storage_type(), clone_from->storage_type());
+      for(size_t i = 0, n = clone_from->aux_shape_count(); i < n; ++i) {
+        TShape ashape = dest->aux_shape(i);
+        CHECK(shape_assign(&ashape, clone_from->aux_shape(i)));
+        dest->CheckAndAllocAuxData(i, ashape);
+      }
+      DCHECK_EQ(dest->aux_shape_count(), clone_from->aux_shape_count());
+    } else {
+      for (size_t i = 0, n = dest->aux_shape_count(); i < n; ++i) {
+        dest->CheckAndAllocAuxData(i, dest->aux_shape(i));
+      }
+      dest->CheckAndAllocData(dest->storage_shape());
+    }
+  }
+
+  /*! \brief Copy the geometry-related blobs (row sparse indexes, etc.) */
+  template<typename xpu>
+  static inline void CopyGeometryBlobs(mshadow::Stream<xpu> *s,
+                                       const NDArray *dest,
+                                       const OpReqType reqi,
+                                       const NDArray &src) {
+    CHECK_EQ(src.aux_shape_count(), dest->aux_shape_count());
+    // My assumption is that the geometry blobs are not large enough to justify an omp loop here,
+    // since the thread synchronization calls for each fork will take longer
+    // than copying a few floats
+    for(size_t i = 0, n = src.aux_shape_count(); i < n; ++i) {
+      const TBlob src_blob = src.aux_data(i);
+      const TBlob dest_blob = dest->aux_data(i);
+      CopyBlob<xpu>(s, dest_blob, reqi, src_blob);
+    }
+  }
+
+  /*! \brief Generic copy NDArray */
+  template<typename xpu>
+  static inline void CopyNDArray(mshadow::Stream<xpu> *s,
+                                 const NDArray& dest,
+                                 const OpReqType reqi,
+                                 const NDArray& src) {
+    DCHECK_NE(dest.storage_type(), kDefaultStorage);
+    DCHECK_EQ(dest.storage_type(), src.storage_type());
+    AllocateGeometry(&dest, &src);
+    CopyGeometryBlobs(s, &dest, reqi, src);
+    CopyBlob(s, dest.data(), reqi, src.data());
+  }
+
+  /*! \brief Get NDArray's data blob, possibly reshaped if necessary to reflect actual
+   *         number of stored items */
+  static inline TBlob GetReshapedBlob(const NDArray& arr) {
```

Comment:
Why do you need this? Isn't the shape already set? 
https://github.com/eric-haibin-lin/mxnet/blob/master/include/mxnet/ndarray.h#L771